mpmath/tests/test_format.py

```diff
@@ -0,0 +1,577 @@
+    assert f'{mp.mpf(0.01):.2g}' == '0.01'
+    assert f'{mp.mpf(0.001):.2g}' == '0.001'
+    assert f'{mp.mpf(1e-4):.2g}' == '0.0001'
+    assert f'{mp.mpf(1e-5):.2g}' == '1e-05'
+    assert f'{mp.mpf(1e-6):.2g}' == '1e-06'
+    assert f'{mp.mpf(1234):.2g}' == '1.2e+03'
+    assert f'{mp.mpf(123):.2g}' == '1.2e+02'
+    assert f'{mp.mpf(12.3):.2g}' == '12'
+    assert f'{mp.mpf(1.23):.2g}' == '1.2'
+    assert f'{mp.mpf(0.123):.2g}' == '0.12'
+    assert f'{mp.mpf(0.0123):.2g}' == '0.012'
+    assert f'{mp.mpf(0.00123):.2g}' == '0.0012'
+    assert f'{mp.mpf(0.000123):.2g}' == '0.00012'
+    assert f'{mp.mpf(0.0000123):.2g}' == '1.2e-05'
+
+    # bad cases from http://bugs.python.org/issue9980
+    assert f'{mp.mpf(38210.0):.12g}' == '38210'
+    assert f'{mp.mpf(37210.0):.12g}' == '37210'
+    assert f'{mp.mpf(36210.0):.12g}' == '36210'
+
+    # alternate g formatting:  always include decimal point and
+    # exactly <precision> significant digits.
+    assert f'{mp.mpf(0):#.0g}' == '0.'
+    assert f'{mp.mpf(0):#.1g}' == '0.'
+    assert f'{mp.mpf(0):#.2g}' == '0.0'
+    assert f'{mp.mpf(0):#.3g}' == '0.00'
+    assert f'{mp.mpf(0):#.4g}' == '0.000'
+
+    assert f'{mp.mpf(0.2):#.0g}' == '0.2'
+    assert f'{mp.mpf(0.2):#.1g}' == '0.2'
+    assert f'{mp.mpf(0.2):#.2g}' == '0.20'
+    assert f'{mp.mpf(0.2):#.3g}' == '0.200'
+    assert f'{mp.mpf(0.2):#.4g}' == '0.2000'
+    assert f'{mp.mpf(0.2):#.10g}' == '0.2000000000'
+
+    assert f'{mp.mpf(2):#.0g}' == '2.'
+    assert f'{mp.mpf(2):#.1g}' == '2.'
+    assert f'{mp.mpf(2):#.2g}' == '2.0'
+    assert f'{mp.mpf(2):#.3g}' == '2.00'
+    assert f'{mp.mpf(2):#.4g}' == '2.000'
+
+    assert f'{mp.mpf(20):#.0g}' == '2.e+01'
+    assert f'{mp.mpf(20):#.1g}' == '2.e+01'
+    assert f'{mp.mpf(20):#.2g}' == '20.'
+    assert f'{mp.mpf(20):#.3g}' == '20.0'
+    assert f'{mp.mpf(20):#.4g}' == '20.00'
+
+    assert f'{mp.mpf(234.56):#.0g}' == '2.e+02'
+    assert f'{mp.mpf(234.56):#.1g}' == '2.e+02'
+    assert f'{mp.mpf(234.56):#.2g}' == '2.3e+02'
+    assert f'{mp.mpf(234.56):#.3g}' == '235.'
+    assert f'{mp.mpf(234.56):#.4g}' == '234.6'
+    assert f'{mp.mpf(234.56):#.5g}' == '234.56'
+    assert f'{mp.mpf(234.56):#.6g}' == '234.560'
+
+
+def test_mpf_float():
+    '''
+    These are additional random tests that check that mp.mpf and fp.mpf yield
+    the same results for default precision.
+    '''
+
+    for _ in range(10000):
+        fmt_str = random_fmt()
+        num = random.choice(
+                (
+                    random.uniform(-1e300, 1e300),
+                    random.uniform(-1e-300, 1e-300)
+                    )
+                )
+
+        assert fmt_str.format(fp.mpf(num)) == fmt_str.format(mp.mpf(num))
+
+    # These additional tests test for mpf initialized form subnormal floats
+    for _ in range(10000):
+        fmt_str = random_fmt()
+        num = fp.mpf('{:.15f}e{:d}'.format(
+                     random.uniform(1, 10), random.randint(-324, -308)))
```

Comment:
If I use something like `random.uniform(0, 1e-308)`, it will yield most numbers with exponent between -308 and -309. This way every subnormal exponent has the same probability to appear.